In [2]:
# Import libraries

import pandas as pd
import nltk

from nltk.sentiment.vader import SentimentIntensityAnalyzer
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer

# import nltk
# import ssl

# try:
#     _create_unverified_https_context = ssl._create_unverified_context
# except AttributeError:
#     pass
# else:
#     ssl._create_default_https_context = _create_unverified_https_context

# nltk.download('all')

In [9]:
df = pd.read_csv('hotel-reviews.csv', encoding = 'UTF-8')
df

User_ID                                        Description  \
0    id10326  The room was kind of clean but had a VERY stro...   
1    id10327  I stayed at the Crown Plaza April -- - April -...   
2    id10328  I booked this hotel through Hotwire at the low...   
3    id10329  Stayed here with husband and sons on the way t...   
4    id10330  My girlfriends and I stayed here to celebrate ...   
..       ...                                                ...   
494  id10820  We stayed at the JW Marriott as a family a wee...   
495  id10821  Hotel is modern and attractive. Convenient to ...   
496  id10822  We were there in October at an unseasonally wa...   
497  id10823  We only spent one day & night and it was amazi...   
498  id10824  The property itself is clean, well-presented a...   

          Browser_Used Device_Used  Positive  
0                 Edge      Mobile         0  
1    Internet Explorer      Mobile         0  
2              Mozilla      Tablet         0  
3     InternetExplorer     Desktop         1  
4                 Edge      Tablet         0  
..                 ...         ...       ...  
494      Google Chrome      Mobile         1  
495      Google Chrome     Desktop         1  
496   InternetExplorer     Desktop         1  
497               Edge      Mobile         1  
498  Internet Explorer      Mobile         0  

[499 rows x 5 columns]

In [10]:
# Create the preprocess_text function

def preprocess_text(text):
    
    # Tokenize the text
    tokens = word_tokenize(text.lower())

    # Remove stop words
    filtered_tokens = [token for token in tokens if token not in stopwords.words('english')]

    # Lemmatize the tokens
    lemmatizer = WordNetLemmatizer()

    lemmatized_tokens = [lemmatizer.lemmatize(token) for token in filtered_tokens]

    # Join the tokens back into a string
    processed_text = ' '.join(lemmatized_tokens)

    return processed_text

df['Description'] = df['Description'].apply(preprocess_text)[:10]
df

User_ID                                        Description  \
0    id10326  room kind clean strong smell dog . generally a...   
1    id10327  stayed crown plaza april -- - april -- , -- --...   
2    id10328  booked hotel hotwire lowest price could find ....   
3    id10329  stayed husband son way alaska cruise . loved h...   
4    id10330  girlfriend stayed celebrate -- th birthday . p...   
..       ...                                                ...   
494  id10820                                                NaN   
495  id10821                                                NaN   
496  id10822                                                NaN   
497  id10823                                                NaN   
498  id10824                                                NaN   

          Browser_Used Device_Used  Positive  
0                 Edge      Mobile         0  
1    Internet Explorer      Mobile         0  
2              Mozilla      Tablet         0  
3     InternetExplorer     Desktop         1  
4                 Edge      Tablet         0  
..                 ...         ...       ...  
494      Google Chrome      Mobile         1  
495      Google Chrome     Desktop         1  
496   InternetExplorer     Desktop         1  
497               Edge      Mobile         1  
498  Internet Explorer      Mobile         0  

[499 rows x 5 columns]

In [19]:
# Initialize NLTK sentiment analyzer
analyzer = SentimentIntensityAnalyzer()

# Create the get_sentiment function
def get_sentiment(text):
    if isinstance(text, str):
        scores = analyzer.polarity_scores(text)
        sentiment = 1 if scores['pos'] > 0 else 0
        return sentiment
    else:
        # Handle non-string inputs (e.g., floats) gracefully
        return None

# Apply get_sentiment function
df['sentiment'] = df['Description'].apply(get_sentiment)
df

User_ID                                        Description  \
0    id10326  room kind clean strong smell dog . generally a...   
1    id10327  stayed crown plaza april -- - april -- , -- --...   
2    id10328  booked hotel hotwire lowest price could find ....   
3    id10329  stayed husband son way alaska cruise . loved h...   
4    id10330  girlfriend stayed celebrate -- th birthday . p...   
..       ...                                                ...   
494  id10820                                                  1   
495  id10821                                                  1   
496  id10822                                                  1   
497  id10823                                                  1   
498  id10824                                                  1   

          Browser_Used Device_Used  Positive  sentiment  
0                 Edge      Mobile         0        1.0  
1    Internet Explorer      Mobile         0        1.0  
2              Mozilla      Tablet         0        1.0  
3     InternetExplorer     Desktop         1        1.0  
4                 Edge      Tablet         0        1.0  
..                 ...         ...       ...        ...  
494      Google Chrome      Mobile         1        NaN  
495      Google Chrome     Desktop         1        NaN  
496   InternetExplorer     Desktop         1        NaN  
497               Edge      Mobile         1        NaN  
498  Internet Explorer      Mobile         0        NaN  

[499 rows x 6 columns]

In [15]:
from sklearn.metrics import classification_report

print(classification_report(df['Positive'], df['sentiment']))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00       169
           1       0.66      1.00      0.80       330

    accuracy                           0.66       499
   macro avg       0.33      0.50      0.40       499
weighted avg       0.44      0.66      0.53       499



/Volumes/MEDIA/[02] PROJECTS/Coding/Data Science/sa-hotel-reviews/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Volumes/MEDIA/[02] PROJECTS/Coding/Data Science/sa-hotel-reviews/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Volumes/MEDIA/[02] PROJECTS/Coding/Data Science/sa-hotel-reviews/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_

## Summary
As you can see, the overall accuracy of this rule-based sentiment analysis model is 66%. Further model training and experimentation could achieve a more desired level of performace.